### PR TITLE
use correct initialisation in linear module

### PR DESCRIPTION
**Pull request by mrTsjolder**
From [this discussion](https://discuss.pytorch.org/t/whats-the-default-initialization-methods-for-layers/3157/17) it appears that the linear layer should be initialised according to (LeCun et al., 1998). However, LeCun et al. promote a variance of `1/n`, rather than a uniform distribution with bounds `1/n`! Also biases are generally expected to be initialised with zeros.

### CODE REVIEW

**Comment by ezyang** [comment]
I'm going to say that we are not going to change this. If you want to submit a doc patch suggesting the alternate initialization method, I'd be very happy to accept it.

**Comment by SsnL** [comment]
I’d say we should initialize bias to zeros. It is very standard and reasonable. We did a similar patch for BN recently.

File: torch/nn/modules/linear.py [hunk 1]
```diff
@@ -53,11 +51,11 @@ def __init__(self, in_features, out_features, bias=True):
         self.reset_parameters()
 
     def reset_parameters(self):
-        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
+        std = 1 / self.in_features ** 0.5
+        bound = std * 3 ** 0.5
+        init.uniform_(self.weight, -bound, bound)
```
Comment (fmassa) [inline torch/nn/modules/linear.py]: Can you use instead `kaiming_uniform_` with a potential different `a`? This is very similar to what we had before, the difference is that there is a `math.sqrt(3)` factor difference, see https://github.com/pytorch/pytorch/commit/f9a99d5504818e2d8211dd9229665fb732a9c61e#diff-531f4c06f42260d699f43dabdf741b6dR55

File: torch/nn/modules/linear.py [hunk 2]
```diff
@@ -120,10 +116,11 @@ def __init__(self, in1_features, in2_features, out_features, bias=True):
         self.reset_parameters()
 
     def reset_parameters(self):
-        bound = 1 / math.sqrt(self.weight.size(1))
+        std = 1 / math.sqrt(self.in1_features * self.in2_features)
```
Comment (fmassa) [inline torch/nn/modules/linear.py]: Why did you change the initialization for the bilinear, is there a reference for it?

**Review by fmassa — COMMENTED** [review]
I have mixed feelings about changing the initialization. This is a backward-incompatible change and people's code might suffer performance change, which will be hard to catch.

File: torch/nn/modules/linear.py (hunk 1 above)
Comment (mrTsjolder) [reply torch/nn/modules/linear.py]:
```suggestion
        init.kaiming_uniform_(self.weight, nonlinearity='linear')
```
would be the more intuitive way to use `kaiming_uniform` (rather than with the leaky relu nonlinearity).

File: torch/nn/modules/linear.py (hunk 2 above)
Comment (mrTsjolder) [reply torch/nn/modules/linear.py]: There's no reference for this, but it is the same math as for the regular linear layer. if you have two standard Gaussian signals, then this initialisation should make the output signal a standard Gaussian as well. 

The main point, however, is still that the desired standard deviation is used as a bound for the uniform, which results in "less desired" standard deviation of the generated samples!

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit f76098a:


* **1/1** broken upstream at merge base 348d421 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=348d42114ea8b5c85288a0f6d80789f97089f29c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f76098ae8adf01ed30b6bcbca610fb5fedb65642).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=77905428)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
